### PR TITLE
Add support for one more hops upgrade

### PR DESCRIPTION
**Pull request by shellyyang1989**
The change:
1. iterates upgrade on all the targets
2. adds an upgrade test from 4.9 to 4.10 to 4.11 for the verification

### CODE REVIEW

File: ci-operator/config/openshift/verification-tests/openshift-verification-tests-master__nightly-4.11-upgrade-from-stable-4.10-from-stable-4.9.yaml [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+      BASE_DOMAIN: qe.devcluster.openshift.com
+      UPGRADE_CHECK_RUN_TAGS: '@upgrade-check and @4.10 and @aws-ipi'
+      UPGRADE_PRE_RUN_TAGS: '@upgrade-prepare and @4.10 and @aws-ipi'
+    post:
+    - chain: cucushift-installer-rehearse-aws-ipi-deprovision
+    pre:
+    - chain: cucushift-installer-rehearse-aws-ipi-provision
+    test:
+    - as: set-upgrade-releases
+      commands: echo "${RELEASE_IMAGE_INTERMEDIATE},${RELEASE_IMAGE_TARGET}" > ${SHARED_DIR}/override-upgrade
```
Comment (jianlinliu) [inline ci-operator/config/openshift/verification-tests/openshift-verification-tests-master__nightly-4.11-upgrade-from-stable-4.10-from-stable-4.9.yaml:53]: How about renaming `override-upgrade` to `upgrade-path`

Comment (shellyyang1989) [reply ci-operator/config/openshift/verification-tests/openshift-verification-tests-master__nightly-4.11-upgrade-from-stable-4.10-from-stable-4.9.yaml:53]: Sure. Let me update. Thanks!

File: ci-operator/step-registry/cucushift/upgrade/toimage/cucushift-upgrade-toimage-commands.sh [hunk 2]
```diff
@@ -100,46 +107,41 @@ function check_history() {
...
-    FORCE_UPDATE=true
-else
-    admin_ack
-fi
+    FORCE_UPDATE=false
+    if ! check_signed; then
+        echo "You're updating to an unsigned images, you must override the verification using --force flag"
+        FORCE_UPDATE=true
+    else
+        admin_ack
```
Comment (jianlinliu) [inline ci-operator/step-registry/cucushift/upgrade/toimage/cucushift-upgrade-toimage-commands.sh:141]: I think `admin_ack` is not on the same level as `check_signed`, if payload is not signed, allow need `FORCE_UPDATE`,   whatever it is already `admin_ack` or not.

Comment (jianlinliu) [reply ci-operator/step-registry/cucushift/upgrade/toimage/cucushift-upgrade-toimage-commands.sh:141]: How about this, just for easy understanding for the logic.
```
    FORCE_UPDATE=false
    if ! check_signed; then
        echo "You're updating to an unsigned images, you must override the verification using --force flag"
        FORCE_UPDATE=true
    end

    if [[ "${FORCE_UPDATE}" == "false" ]]
        admin_ack
    fi
```

Or even we can move `FORCE_UPDATE` judgement into `admin_ack` function?

File: ci-operator/step-registry/cucushift/upgrade/toimage/cucushift-upgrade-toimage-commands.sh [hunk 3]
```diff
@@ -100,46 +107,41 @@ function check_history() {
...
+    echo "RELEASE_IMAGE_TARGET is ${RELEASE_IMAGE_TARGET}"
 
-echo -e "RELEASE_IMAGE_LATEST release info:\n"
-oc adm release info "${RELEASE_IMAGE_LATEST}"
+    echo -e "Current client/server version: oc version\n"
+    oc version
 
-echo -e "RELEASE_IMAGE_TARGET release info:\n"
-oc adm release info "${RELEASE_IMAGE_TARGET}"
+    source_minor_version="$(oc get clusterversion --no-headers | awk '{print $2}' | cut -f2 -d.)"
```
Comment (jianlinliu) [inline ci-operator/step-registry/cucushift/upgrade/toimage/cucushift-upgrade-toimage-commands.sh:129]: if vars (e.g: `source_minor_version`) will be referenced in the following function, I suggest always `export` it for safety.

File: ci-operator/step-registry/cucushift/upgrade/toimage/cucushift-upgrade-toimage-commands.sh (hunk 2 above)
Comment (shellyyang1989) [reply ci-operator/step-registry/cucushift/upgrade/toimage/cucushift-upgrade-toimage-commands.sh:141]: Thanks for the suggestion! Let me update.

File: ci-operator/step-registry/cucushift/upgrade/toimage/cucushift-upgrade-toimage-commands.sh [hunk 4]
```diff
@@ -100,46 +107,41 @@ function check_history() {
...
+    else
+        admin_ack
+    fi
 
-upgrade 
-check_upgrade_status 
-check_history 
+    upgrade 
+    check_upgrade_status 
+    check_history
```
Comment (jiajliu) [inline ci-operator/step-registry/cucushift/upgrade/toimage/cucushift-upgrade-toimage-commands.sh:146]: How to deal with the unready status between 1st update and 2nd update? check_upgrade_status seems not enough to cover the unready status we hit in upgrade ci. For example: mcp is updating and upgradeable=false.

File: ci-operator/step-registry/cucushift/upgrade/toimage/cucushift-upgrade-toimage-commands.sh (hunk 3 above)
Comment (shellyyang1989) [reply ci-operator/step-registry/cucushift/upgrade/toimage/cucushift-upgrade-toimage-commands.sh:129]: Thanks for the suggestion. Wrt the best practice, i'd like to seek your expertise
1) if a var is readonly to the functions and subprocesses, making it an env var better?
2) if a var is writable to all the functions, making it a global var better (by default it's a global var)?

File: ci-operator/step-registry/cucushift/upgrade/toimage/cucushift-upgrade-toimage-commands.sh (hunk 4 above)
Comment (shellyyang1989) [reply ci-operator/step-registry/cucushift/upgrade/toimage/cucushift-upgrade-toimage-commands.sh:146]: yeah. I should have moved the health check in, but i forgot it. Let me update. Thanks!

File: ci-operator/step-registry/cucushift/upgrade/toimage/cucushift-upgrade-toimage-commands.sh (hunk 3 above)
Comment (jianlinliu) [reply ci-operator/step-registry/cucushift/upgrade/toimage/cucushift-upgrade-toimage-commands.sh:129]: Yeah, you are right. Subprocesses can not write a env inherited from its parent process.

**Comment by jiajliu** [comment]
I see job `ci/rehearse/periodic-ci-openshift-verification-tests-master-nightly-4.11-upgrade-from-stable-4.10-from-stable-4.9-aws-ipi — Job succeed` succeed. But from the build log, the steps in workflow `cucushift-upgrade` seems skipped.
```
...
INFO[2022-06-09T04:39:01Z] Running multi-stage phase test               
INFO[2022-06-09T04:39:01Z] Running step aws-ipi-set-upgrade-releases.   
INFO[2022-06-09T04:[39](https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_release/29255/rehearse-29255-periodic-ci-openshift-verification-tests-master-nightly-4.11-upgrade-from-stable-4.10-from-stable-4.9-aws-ipi/1534748171660431360#1:build-log.txt%3A39):22Z] Step aws-ipi-set-upgrade-releases succeeded after 20s. 
INFO[2022-06-09T04:39:22Z] Step phase test succeeded after 20s.         
INFO[2022-06-09T04:39:22Z] Running multi-stage phase post
...
```

**Comment by shellyyang1989** [comment]
> I see [job](https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_release/29255/rehearse-29255-periodic-ci-openshift-verification-tests-master-nightly-4.11-upgrade-from-stable-4.10-from-stable-4.9-aws-ipi/1534748171660431360) `ci/rehearse/periodic-ci-openshift-verification-tests-master-nightly-4.11-upgrade-from-stable-4.10-from-stable-4.9-aws-ipi — Job succeed` succeed. But from the build log, the steps in workflow `cucushift-upgrade` seems skipped.

Right. It's interesting. Let me trigger a test on it.

**Comment by shellyyang1989** [comment]
> > I see [job](https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_release/29255/rehearse-29255-periodic-ci-openshift-verification-tests-master-nightly-4.11-upgrade-from-stable-4.10-from-stable-4.9-aws-ipi/1534748171660431360) `ci/rehearse/periodic-ci-openshift-verification-tests-master-nightly-4.11-upgrade-from-stable-4.10-from-stable-4.9-aws-ipi — Job succeed` succeed. But from the build log, the steps in workflow `cucushift-upgrade` seems skipped.
> 
> Right. It's interesting. Let me trigger a test on it.

Well, the test defined in the job overrides the one in workflow. So workflow is skipped. Let me update.

**Comment by shellyyang1989** [comment]
The 4.9->4.10->4.11 upgrade [test](https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_release/29255/rehearse-29255-periodic-ci-openshift-verification-tests-master-nightly-4.11-upgrade-from-stable-4.10-from-stable-4.9-aws-ipi/1535081440964775936) passed.

But oc version is not consistent with the target image. Needs update.

**Comment by shellyyang1989** [comment]
cc @liangxia @jhou1 for review because there are some changes on the upgrade workflow and test config. Thanks!

File: ci-operator/config/openshift/verification-tests/openshift-verification-tests-master__nightly-4.11-upgrade-from-stable-4.10-from-stable-4.9.yaml [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+      memory: 200Mi
+tests:
+- as: aws-ipi
+  cron: 0 18 */3 * *
+  steps:
+    cluster_profile: aws-qe
+    env:
+      BASE_DOMAIN: qe.devcluster.openshift.com
+      UPGRADE_CHECK_RUN_TAGS: '@upgrade-check and @4.10 and @aws-ipi'
+      UPGRADE_PRE_RUN_TAGS: '@upgrade-prepare and @4.10 and @aws-ipi'
```
Comment (liangxia) [inline ci-operator/config/openshift/verification-tests/openshift-verification-tests-master__nightly-4.11-upgrade-from-stable-4.10-from-stable-4.9.yaml:46]: Shall we use 4.9 here ?

Comment (liangxia) [reply ci-operator/config/openshift/verification-tests/openshift-verification-tests-master__nightly-4.11-upgrade-from-stable-4.10-from-stable-4.9.yaml:46]: I think the installation use stable 4.9 payload, so we'd expect the upgrade prepare choosing 4.9 test scenarios. Correct me if I got that wrong.

Comment (shellyyang1989) [reply ci-operator/config/openshift/verification-tests/openshift-verification-tests-master__nightly-4.11-upgrade-from-stable-4.10-from-stable-4.9.yaml:46]: yeah, right.

Comment (shellyyang1989) [reply ci-operator/config/openshift/verification-tests/openshift-verification-tests-master__nightly-4.11-upgrade-from-stable-4.10-from-stable-4.9.yaml]: Updated. Thanks!

**Comment by jianping-shu** [comment]
This might be normal, which is usually needed for CCO mode=Manual.
The user needs to create the new credentials for the upgrade release manually and add Upgradeable annotation to inform CCO module. 



> ```
> 2022-06-09T08:03:02.938118156Z I0609 08:03:02.938016       1 event.go:285] Event(v1.ObjectReference{Kind:"ClusterVersion", Namespace:"openshift-cluster-version", Name:"version", UID:"", APIVersion:"config.openshift.io/v1", ResourceVersion:"", FieldPath:""}): type: 'Warning' reason: 'PreconditionBlock' Preconditions failed for payload loaded version="4.11.0-0.nightly-2022-06-06-201913" image="registry.build01.ci.openshift.org/ci-op-vtiir1yc/release@sha256:49af2f8eeef5a24b2418aa6ba0be097a6e74bc747d35d403afe51ff1b173fa0b": Precondition "ClusterVersionUpgradeable" failed because of "MissingUpgradeableAnnotation": Cluster operator cloud-credential should not be upgraded between minor versions: Upgradeable annotation cloudcredential.openshift.io/upgradeable-to on cloudcredential.operator.openshift.io/cluster object needs updating before upgrade. See Manually Creating IAM documentation for instructions on preparing a cluster for upgrade.
> ```
> 
> The [job](https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_release/29255/rehearse-29255-periodic-ci-openshift-verification-tests-master-nightly-4.11-upgrade-from-stable-4.10-aws-ipi-disconnected-private/1534748171811426304) ran into the cco error. cc @jianping-shu for awareness.

**Comment by shellyyang1989** [comment]
> This might be normal, which is usually needed for CCO mode=Manual. The user needs to create the new credentials for the upgrade release manually and add Upgradeable annotation to inform CCO module.
> 
> > ```
> > 2022-06-09T08:03:02.938118156Z I0609 08:03:02.938016       1 event.go:285] Event(v1.ObjectReference{Kind:"ClusterVersion", Namespace:"openshift-cluster-version", Name:"version", UID:"", APIVersion:"config.openshift.io/v1", ResourceVersion:"", FieldPath:""}): type: 'Warning' reason: 'PreconditionBlock' Preconditions failed for payload loaded version="4.11.0-0.nightly-2022-06-06-201913" image="registry.build01.ci.openshift.org/ci-op-vtiir1yc/release@sha256:49af2f8eeef5a24b2418aa6ba0be097a6e74bc747d35d403afe51ff1b173fa0b": Precondition "ClusterVersionUpgradeable" failed because of "MissingUpgradeableAnnotation": Cluster operator cloud-credential should not be upgraded between minor versions: Upgradeable annotation cloudcredential.openshift.io/upgradeable-to on cloudcredential.operator.openshift.io/cluster object needs updating before upgrade. See Manually Creating IAM documentation for instructions on preparing a cluster for upgrade.
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > The [job](https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_release/29255/rehearse-29255-periodic-ci-openshift-verification-tests-master-nightly-4.11-upgrade-from-stable-4.10-aws-ipi-disconnected-private/1534748171811426304) ran into the cco error. cc @jianping-shu for awareness.

Thanks!

I see it's Manual Static CCO in the profile. Does it mean that in the mode we need to do extra configuration before upgrade? Also would you mind directing me the the doc or case for the detailed procedures?

**Comment by jianping-shu** [comment]
Here is the link
https://docs.openshift.com/container-platform/4.10/updating/updating-cluster-within-minor.html#manually-maintained-credentials-upgrade_updating-cluster-within-minor

In short, if there's new CredentialsRequest introduced between old and new releases, then need to create the Secret for the CredentialsRequest and add upgradeable-to annotation.
If no new CredentialsRequest, add upgradeable-to annotation directly.

> I see it's Manual Static CCO in the profile. Does it mean that in the mode we need to do extra configuration before upgrade? Also would you mind directing me the the doc or case for the detailed procedures?

**Comment by shellyyang1989** [comment]
The multi-hop upgrade passed. But upgrade for disconnected cluster failed due to CCO issue. Let me fix.

**Comment by shellyyang1989** [comment]
> Here is the link https://docs.openshift.com/container-platform/4.10/updating/updating-cluster-within-minor.html#manually-maintained-credentials-upgrade_updating-cluster-within-minor
> 
> In short, if there's new CredentialsRequest introduced between old and new releases, then need to create the Secret for the CredentialsRequest and add upgradeable-to annotation. If no new CredentialsRequest, add upgradeable-to annotation directly.
> 
> > I see it's Manual Static CCO in the profile. Does it mean that in the mode we need to do extra configuration before upgrade? Also would you mind directing me the the doc or case for the detailed procedures?

Thanks!

Per the doc, seems no extra credentials are introduced in 4.10. Could you please help confirm if only upgradeable-to annotation is required in the 4.9->4.10->4.11 upgrade?

**Comment by shellyyang1989** [comment]
Talked with CCO folks, manual static CCO is used by default in aws disconnected cluster. In the mode, we need to create IAM, secret, annotation as needs before upgrade. So to avoid failing upgrade job for disconnected aws clusters, removed the job aws-ipi-disconnected-private at the moment. I'll open a separate pr to support it later.

**Comment by shellyyang1989** [comment]
The tests failed at installation. Let's retest.

/retest

**Comment by shellyyang1989** [comment]
Hi @liangxia the job ci/rehearse/periodic-ci-openshift-verification-tests-master-nightly-4.11-upgrade-from-stable-4.10-aws-ipi-disconnected-private has been dropped in the PR, but it's always displayed in the failed test list. Should I do anything to it? Will it block the merge?

**Comment by shellyyang1989** [comment]
@jianlinliu @jiajliu @liangxia could anyone take a look when you have time? Thanks!

**Pull request merged**